lib/execution.go

```diff
@@ -0,0 +1,562 @@
+	}
+
+	close(es.resumeNotify)
+
+	return nil
+}
+
+// ResumeNotify returns a channel which will be closed (i.e. could
+// be read from) as soon as the test execution is resumed.
+//
+// Since tests would likely be paused only rarely, unless you
+// directly need to be notified via a channel that the test
+// isn't paused or that it has resumed, it's probably a good
+// idea to first use the IsPaused() method, since it will be much
+// faster.
+//
+// And, since tests won't be paused most of the time, it's
+// probably better to check for that like this:
+//   if executionState.IsPaused() {
+//       <-executionState.ResumeNotify()
+//   }
+func (es *ExecutionState) ResumeNotify() <-chan struct{} {
+	es.pauseStateLock.RLock()
+	defer es.pauseStateLock.RUnlock()
+	return es.resumeNotify
+}
+
+// GetPlannedVU tries to get a pre-initialized VU from the buffer channel. This
+// shouldn't fail and should generally be an instantaneous action, but if it
+// doesn't happen for MaxTimeToWaitForPlannedVU, a warning will be printed. If
+// we reach that timeout more than MaxRetriesGetPlannedVU number of times, this
+// function will return an error, since we either have a bug with some
+// executor, or the machine is very, very overloaded.
+func (es *ExecutionState) GetPlannedVU(logger *logrus.Entry, modifyAtiveVUCount bool) (VU, error) {
+	for i := 1; i <= MaxRetriesGetPlannedVU; i++ {
+		select {
+		case vu := <-es.vus:
+			if modifyAtiveVUCount {
+				es.ModCurrentlyActiveVUsCount(+1)
+			}
+			//TODO: set environment and exec
+			return vu, nil
+		case <-time.After(MaxTimeToWaitForPlannedVU):
+			logger.Warnf("Could not get a VU from the buffer for %s", time.Duration(i)*MaxTimeToWaitForPlannedVU)
+		}
+	}
+	return nil, fmt.Errorf(
+		"could not get a VU from the buffer in %s",
+		MaxRetriesGetPlannedVU*MaxTimeToWaitForPlannedVU,
+	)
+}
+
+// SetInitVUFunc is called by the execution scheduler's init function, and it's
+// used for setting the "constructor" function used for the initializing
+// unplanned VUs.
+//
+// TODO: figure out a better dependency injection method?
+func (es *ExecutionState) SetInitVUFunc(initVUFunc InitVUFunc) {
+	es.initVUFunc = initVUFunc
+}
+
+// GetUnplannedVU checks if any unplanned VUs remain to be initialized, and if
+// they do, it initializes one and returns it. If all unplanned VUs have already
+// been initialized, it returns one from the global vus buffer.
+//
+// IMPORTANT: GetUnplannedVU() doesn't do any checking if the requesting
+// executor is actually allowed to have the VU at this particular time.
+// Executors are trusted to correctly declare their needs (via their
+// GetExecutionRequirements() methods) and then to never ask for more VUs than
+// they have specified in those requirements.
+func (es *ExecutionState) GetUnplannedVU(ctx context.Context, logger *logrus.Entry) (VU, error) {
+	remVUs := atomic.AddInt64(es.uninitializedUnplannedVUs, -1)
+	if remVUs < 0 {
```

Comment:
Would it be safer to do a check here instead and only decrement it right before the VU has been initialized? E.g.:

```golang
if atomic.LoadInt64(es.uninitializedUnplannedVUs) == 0 {
	...
	return err
}
...
remVUs := atomic.AddInt64(es.uninitializedUnplannedVUs, -1)
vu, err := es.InitializeNewVU(ctx, logger)
...
return vu, err
```

This would avoid the `atomic.AddInt64(es.uninitializedUnplannedVUs, 1)` call.

Reply:
Yeah, but than there could be two gorutines which  load it and than add it thinking they were the only ones .. which would be a bug 

To be honest I don't like the whole initialization and adding/returning thing here, but couldn't figure out how to refactor it and I hope I will get some ideas while writing tests, because there .. it is the primary source of lines of code :(

Reply:
Hhmm right, the `< 0` check is needed. Nevermind.